### PR TITLE
feat(classifier): SVM pre-filter ahead of the LLM

### LINKED ISSUE #45: Two-stage pipeline: all-minilm pre-filter + LLM verify
## Goal

Cut Ollama load on the Pi by ~97% without sacrificing precision, using the insight from #36's experiment results.

## Background

The #36 experiment (PR #43, closed without merge) showed:

- All-minilm (46 MB) + LinearSVC with class_weight=\"balanced\" reaches **88% recall at ~97% precision on the negative class**. In other words: it correctly rejects ~97% of non-SWE titles with very few false negatives.
- Pareto-dominated by qwen2.5:1.5b on precision/recall tradeoffs when used as a sole classifier.
- But as a **pre-filter** stage, it could cut 97% of LLM calls on negatives while still letting qwen2.5:1.5b verify the remaining 3% where precision matters.

## Shape

Pipeline: title -> embed (all-minilm) -> SVM says "maybe SWE" or "definitely not" -> only on "maybe" do we call the qwen2.5:1.5b LLM -> final answer.

Concretely:

1. Ship all-minilm alongside qwen2.5:1.5b in the ollama image (~50 MB extra).
2. Train the SVM head on the #36 ground-truth labels; ship the pickle (\`~15 KB\`) in the repo under \`classifier/svm_model.pkl\` or similar.
3. On classify: embed the title, predict with SVM. If SVM rejects with confidence, return False immediately. Otherwise, fall through to the existing qwen2.5:1.5b + few-shot path.
4. A config toggle (\`CLASSIFIER_PREFILTER_ENABLED\`) so we can A/B or kill-switch in prod without a revert.

Expected effect on Pi: a reclassify pass that currently takes ~25-35 min drops to ~2-5 min because ~97% of calls terminate at SVM. Daily scrapes return results near-instantly.

## Open questions

- **Retraining cadence.** Labels drift as new job titles appear. Monthly? On every release? Triggered manually?
- **Where does training live?** Script in \`scripts/\` that reads the same \`roles-to-label.csv\` format, writes \`classifier/svm_model.pkl\`. Regenerate at build time or commit the pickle?
- **Confidence threshold.** Tune the SVM decision threshold against the labelled set — what's the Pareto-optimal recall the pre-filter should preserve?
- **Fallback.** If all-minilm isn't pulled in Ollama (e.g. fresh install), how do we degrade? Probably: skip pre-filter, go straight to LLM. Log a warning.

## Dependencies / related

- **Built on** #36 experiment results (PR #43 artifacts remain on the branch for reference).
- **Defers** #17 (classifier tuning harness) — if this two-stage approach lands and works, the harness's primary use case (comparing small LLMs) becomes much less urgent.

## Estimated effort

- Training pipeline + pickle generator: 30-60 min.
- Integration into \`src/classifier/client.py\` with config toggle: 30-60 min.
- Eval + tuning + docs: 30-60 min.
- Total: roughly one afternoon.

### PR DESCRIPTION
**Pull request by maciej-makowski**
Closes #45.

## Summary

Two-stage classification:

1. **Stage 1 — SVM pre-filter.** Embed each title via Ollama `all-minilm` (~46 MB on disk, 384-dim), feed through a LinearSVC (class-weight balanced). Titles the SVM confidently classifies as not-SWE short-circuit to \`False\` without any LLM call. Threshold is tuned via 5-fold CV on out-of-fold decision scores to preserve ≥95% positive-class recall.
2. **Stage 2 — LLM confirm.** Whatever the SVM can't confidently reject falls through to the existing \`qwen2.5:1.5b\` / SYSTEM + few-shot chat path. Unchanged from #35.

## Why JSON over pickle

Model shipped at \`classifier/prefilter.json.gz\` (~4 KiB): coefficients, intercept, threshold, metadata. No Python/sklearn version lock. Inference is a pure-Python dot product + threshold — scraper image doesn't need sklearn or numpy. Scikit-learn moved to \`[project.optional-dependencies].training\`.

## Files

- New: \`scripts/train_prefilter.py\`, \`src/classifier/prefilter.py\`, \`classifier/prefilter.json.gz\`, \`classifier/training_data.csv.gz\`, \`tests/integration/test_prefilter.py\`.
- Modified: \`src/classifier/client.py\` (integrate pre-filter, gated on \`settings.classifier_prefilter_enabled\`), \`src/config.py\`, \`Containerfile.ollama\` (bake \`all-minilm\`), \`Makefile\` (\`make retrain-prefilter\`, \`make eval-prefilter\`), \`pyproject.toml\`, docs.

## Kill-switch

\`CLASSIFIER_PREFILTER_ENABLED=false\` env → stage 1 skipped, every title goes to LLM. Same as pre-merge behaviour.

## Graceful degradation

- Embed fails or model file missing → WARN log, every title falls through to LLM. Scrape still completes.
- LLM fails (after pre-filter) → existing behaviour: exception bubbles, scrape retries. Per \"only fail if both failed\" discussion.

## Numbers (5-fold CV on 1381 hand-labels)

| Metric | Pre-filter stage alone | qwen2.5:1.5b baseline (for reference) |
|---|---|---|
| Positive recall | **96.1%** | 72.5% |
| Precision | 16.2% | 69.8% |
| LLM call rate after stage 1 | **21.9%** | (100%) |

The pre-filter's precision is intentionally low — its job is to let through any plausible SWE, not to be the final answer. The LLM does the precision work on the 22% that survive.

## Test plan

- [x] \`uv run pytest tests/unit/ tests/integration/\` — 123 passed (7 new pre-filter tests + 2 new classifier-integration tests for enabled / kill-switch paths).
- [x] \`uv run ruff check\` + \`ruff format --check\` on \`src/ tests/ scripts/\` — clean.
- [x] Pre-commit hooks pass.
- [x] \`make retrain-prefilter\` deterministically produces \`classifier/prefilter.json.gz\`.
- [x] \`make eval-prefilter\` prints CV metrics without rewriting the model.
- [x] Smoke test via \`python -m src.classifier.prefilter\`: confirms \"Research Engineer, Pre-training\" and \"Account Executive, Startups\" short-circuit; \"Senior Software Engineer\" falls through to LLM.
- [x] **Pi reclassify timing** — verified: 1h 55m → 21m (5.5×) on the Pi with CPU-prio profile, same 2086-row DB. See the comparison comment below. — record before (qwen-only) vs after (two-stage) on the same ~2086-row DB. The before timing is currently in-flight from the Pi's current reclassify; I'll add both numbers as a comment on this PR when it lands.

## Follow-ups
- #46 — distill a bigger-LLM oracle to train a higher-quality SVM, potentially drop qwen entirely.

### CODE REVIEW

**Comment by maciej-makowski** [comment]
## Measured on Pi (2086 postings, identical CPU-prio profile)

Both runs reclassify the whole DB on the Pi with the profile from #42 applied (`CPUWeight=10`, `IOWeight=10`, `Nice=15`, `cpu_shares=128` on ollama).

| | Time | Per-title | Predicted SWE |
|---|---:|---:|---:|
| **Before** — qwen2.5:1.5b only, `/api/chat` + few-shot | **1h 55m 02s** | ~3.3 s | 85 |
| **After** — SVM pre-filter + qwen confirm | **20m 57s** | ~0.6 s | 69 |
| **Speedup** | **5.5×** | | –16 |

The count delta (85 → 69) breaks down into two components:

- **Filter recall.** 5-fold CV on the 1381 hand-labels predicted ~96% positive recall, so ~4% of true SWE titles get silently rejected. On 85 predicted SWE, that's ~3 expected misses.
- **Distribution shift for the LLM.** Stage 2 now only sees the SVM's "maybe" bucket (~22% of all titles, with a ~10% base rate of SWE vs ~4% on the full set). qwen's behaviour on this harder distribution looks slightly more conservative — likely why the remaining delta (~13 titles) flipped from "yes" to "no".

Final per-company counts (after):
```
 anthropic  |  6 | 436
 deepmind   |  3 |  88
 meta       | 14 | 611
 openai     | 21 | 643
 perplexity | 15 |  74
 xai        | 10 | 234
 total      | 69 | 2086
```

Deployment verified to go through the template path: live files were regenerated by `make deploy PROFILE=pi` (not `make install-compose`), systemd reports `CPUWeight=10 / IOWeight=10 / Nice=15`, ollama image has both `qwen2.5:1.5b` and `all-minilm` baked in, `/app/classifier/prefilter.json.gz` is present inside the scraper container (requires the Containerfile.scraper fix in commit ad3b573).

Checking the last test-plan box.
